### PR TITLE
ota_client.py: improve _create_regular_files

### PR DESCRIPTION
**Pull request by Bodong-Yang**
In the previous implementation, when we handle the first copy of hardlink file, the main process will be blocked.
In the new implementation, handling first copy of hardlink file is done in the _create_regular_file method, so only a subprocess will be blocked.

More details as follow:
* hardlink_dict is only used in the main process now, previous_reginf is passed instead of the whole dict,
* hardlink_dict is sha256hash[tuple[reginf, event]] now, a **multiprocessing.Event** object is used to ensure the first_copy.
* process pool dispatching logics is placed back to _create_regular_files method.

### CODE REVIEW

**Comment by obi-t4** [comment]
@Bodong-Yang 
Can you revert auto generated files - app2/otaclient*_pb2*.py - before review?

**Comment by Bodong-Yang** [comment]
> @Bodong-Yang Can you revert auto generated files - app2/otaclient*_pb2*.py - before review?

Which commit should I revert to?

**Comment by obi-t4** [comment]
> Which commit should I revert to?

app2/otaclient*_pb2*.py
Your commit changes 4 files. but 3 are unnecessary.

**Comment by Bodong-Yang** [comment]
> > Which commit should I revert to?
> 
> app2/otaclient*_pb2*.py Your commit changes 4 files. but 3 are unnecessary.

Oh that is cause by the git merge, I should replace these files with upstream version instead of merge them!

**Comment by obi-t4** [comment]
> Oh that is cause by the git merge

Your git or editor setting or something?

**Comment by Bodong-Yang** [comment]
> > Oh that is cause by the git merge
> 
> Your git or editor setting or something?

resolved in the latest commit.

File: app2/ota_client.py [hunk 1]
```diff
@@ -370,117 +375,103 @@ def _create_symbolic_links(self, list_file, standby_path):
...
+                    else:
+                        pool.apply_async(
+                            _create_regfile_func,
+                            (reginf,),
+                            error_callback=error_callback,
+                        )
+
                 pool.close()
                 while len(processed_list) < len(reginf_list):
-                    self._regular_files_processed = len(processed_list)  # via setter
```
Comment (obi-t4) [inline app2/ota_client.py:399]: this line is required.

File: app2/ota_client.py [hunk 2]
```diff
@@ -370,117 +375,103 @@ def _create_symbolic_links(self, list_file, standby_path):
...
-                    hardlink_dict,
-                    processed_list,
-                    pool.apply_async,
-                    error_callback,
-                )
+                hardlink_dict = dict()  # sha256hash[tuple[reginf, event]]
+
+                for reginf in reginf_list:
+                    if reginf.nlink >= 2:
+                        first_copy = not reginf.sha256hash in hardlink_dict
```
Comment (obi-t4) [inline app2/ota_client.py:406]: first_copy seems redundant. I think we can remove this line, and we can evaluate `first_copy` boolean value in the function
_create_regular_file:
`first_copy = prev_reginf.path == reginf.path`

File: app2/ota_client.py [hunk 3]
```diff
@@ -370,117 +375,103 @@ def _create_symbolic_links(self, list_file, standby_path):
...
+                        prev_reginf, event = hardlink_dict.setdefault(
+                            reginf.sha256hash, (reginf, manager.Event())
+                        )
+
+                        pool.apply_async(
+                            _create_regfile_func,
+                            (reginf, prev_reginf),
```
Comment (obi-t4) [inline app2/ota_client.py:413]: Can you please add comments? like:
`# the positional argument in a tuple`
`# the keyword arguments in a dictionary`
(Not all of us are familiar with applying_async usage :smiley: )

**Review by obi-t4 — COMMENTED** [review]
I left some comments, but basically, this PR is OK! Thanks!

File: app2/ota_client.py (hunk 1 above)
Comment (Bodong-Yang) [reply app2/ota_client.py:399]: Oh I miss understood the usage of this line, let me fix it.

Comment (obi-t4) [reply app2/ota_client.py:399]: > to keep track of how many files are processed?

Yes!

File: app2/ota_client.py (hunk 2 above)
Comment (Bodong-Yang) [reply app2/ota_client.py]: Yes you are right!

Updated on commit 69cc179.

File: app2/ota_client.py (hunk 3 above)
Comment (Bodong-Yang) [reply app2/ota_client.py:412]: Sure!

**Review by obi-t4 — APPROVED** [review]
完璧 :+1:

**Comment by Bodong-Yang** [comment]
Oh by the way, I also think that we can use multiprocessing to accelerate the processing of regular_file.txt, since it has thousands of records.

**Comment by obi-t4** [comment]
> Oh by the way, I also think that we can use multiprocessing to accelerate the processing of regular_file.txt, since it has thousands of records.

We have to measure it. If the consumption time to process regular_file.txt is a few percent, I think we don't need to optimize it.

**Comment by Bodong-Yang** [comment]
> > Oh by the way, I also think that we can use multiprocessing to accelerate the processing of regular_file.txt, since it has thousands of records.
> 
> We have to measure it. If the consumption time to process regular_file.txt is a few percent, I think we don't need to optimize it.

Actually this optimization makes the code simpler, 
please check the commit 046e40a.

**Comment by obi-t4** [comment]
> please check the commit 046e40a.

sorry new code should be separated PR. Or, we can't close this PR :pray:

**Comment by Bodong-Yang** [comment]
> > please check the commit [046e40a](https://github.com/tier4/ota-client/commit/046e40a85989517c3081ffd05616a7d5581fa035).
> 
> sorry new code should be separated PR. Or, we can't close this PR pray

I think this change can be within this PR since the change is minor actually.
The code do the same thing as before.

File: app2/ota_client.py [hunk 4]
```diff
@@ -370,117 +375,113 @@ def _create_symbolic_links(self, list_file, standby_path):
...
+                            {"hardlink_event": event},
+                            error_callback=error_callback,
+                        )
+                    else:
+                        pool.apply_async(
+                            _create_regfile_func,
+                            (reginf,),
+                            error_callback=error_callback,
+                        )
+                self._total_regular_files = total_regular_files
```
Comment (obi-t4) [inline app2/ota_client.py:431]: `self._total_regular_files is total` should be total number of regular files.

Comment (Bodong-Yang) [reply app2/ota_client.py:431]: Yes, I use this counter variable `total_regular_files ` to count the number of regular files in the for loop.

Comment (Bodong-Yang) [reply app2/ota_client.py:431]: > `self._total_regular_files is total` should be total number of regular files. Not number of regular files processed.

Yes I am doing what you said,
```
for reginf in reginf_list:
    total_regular_files += 1
```
is the same as `len(reginf_list)` in previous implementation.

Comment (Bodong-Yang) [reply app2/ota_client.py:431]: The reason why I count it in the for loop is as follow:
```
                # imap_unordered return a lazy itorator without blocking
                # so we cannot use len() over it
```

**Review by obi-t4 — CHANGES_REQUESTED** [review]
please modify total_regular_files.

File: app2/ota_client.py (hunk 4 above)
Comment (obi-t4) [reply app2/ota_client.py:431]: :thinking: 
previous implementation:
```
self._total_regular_files = len(reginf_list) # via setter
````
I think you're saying `self._regular_files_processed`?

**Comment by Bodong-Yang** [comment]
> please modify total_regular_files.

Fix in commit f0680f4.

**Review by obi-t4 — APPROVED** [review]
LGTM!

**Pull request merged**